### ISSUE TITLE
internal data structure

### ISSUE THREAD

**Issue by mathause** [MESMER-group/mesmer#106]
NOTE: this issue is very much a draft at the moment.

---

We should give some thought on the internal data structure in mesmer. IMHO this is one of the more important but also difficult things to decide on. Generally the idea is to move to a xarray-based data structure. However, it is unclear how we want to carry metadata around (e.g. model name, experiment, etc.).

# Current status

Currently a triply nested dictionary of numpy arrays. Something that roughly looks like:

```python
arr = "array"
data = dict(model=dict(ssp=dict(tas=arr)))
```

# Ideas

## 1. Keep as is

**Pros/ Cons**
- No need to change anything
- Data access is awkward (`data["CESM"]["ssp585"]["tas"]`)
- What if another layer is needed?
- ...

## 2. One huge DataArray

We could potentially create one enormous DataArray that encapsulates all necessary information as coordinates (and non-dimension coordinates). It could look something like this:

```
<xarray.DataArray (time, ens, gp, realisation)>
array([[1, 2, 3]])
Coordinates:
    model    (ens) <U5 'CESM' 'CESM' 'MIROC'
    exp      (ens) <U6 'ssp585' 'ssp126' 'ssp585'
  * time     (time) int32 1
Dimensions without coordinates: ens, gp, realisation
```

**Pros/ Cons**
- A well understood and well tested data structure
- Can get unwieldy if too big
- Difficult to decide where to draw the border
- Need to be careful what becomes a dimension (e.g. not all models have all scenarios, so it cannot be a dimension)
- Cannot have hist and projections in the same array
- Cannot combine different models if they have a different grid

## 3. DataList

_DataList_ is a data structure inspired by [ESMValTool](https://github.com/ESMValGroup/ESMValTool/) that I used extensively for my analysis of the CMIP6 data for IPCC. It is used as a flat data container to store model data (e.g. `xr.DataArray` objects) and its metadata. It is a list of data, metadata tuples:
```python
datalist = [
    (ds, meta),
    (ds, meta),
    ...
]
```
where `ds` is a data store (e.g. an `xr.DataArray`) and meta is a `dict` containing metadata, e.g. `meta = {"model": "CESM", exp: "ssp585", ...}`. This allows to (1) work with a convenient flat data structure (no nested loops), (2) store arbitrary data (e.g. `xr.DataArray` objects with different grids), (3) carry around metadata without having to alter the data store.

_DataList_ structure could be turned into a class, which would allow for a convenient interface. E.g. it could allow for things like

```python
datalist.select(exp="ssp585")
datalist.map(compute_time_average, period=slice(1850, 1900))
```

<details>

An implementation could look like (outside from mesmer):

```python
class DataList:
    
    def __init__(self, data: List[Any], meta: List[dict]):
        
        assert len(data) == len(meta)
        ...

    def select(self, **attributes):
      """Select specific data described by metadata."""

    def remove(self, **attributes):
      """Remove specific data described by metadata."""

    def map(self, func, pass_meta=False, **kwargs):
      """loop over DataList and apply a function"""
    # could also accept function names, e.g.
    # map("sel", time=slice("1850", "1900"))

    def to_dataarray(self, along):
      """concatenate data along a dimension and return as xr.DataArray"""

    def groupby(group):
        "group over an attribute"

def join(*datalists, on=("model", "exp", "ens"), check=True):
    """align datalists (inner join)"""
```
</details>

**Pros/ Cons**
- Flat & flexible data structure
- Not implemented yet (but should be reasonably straightforward - I have most of the code)
- Why two data structures (DataList and combined DataArrays)

## 4. DataTree

WIP implementation of a tree-like hierarchical data structure for xarray, see https://github.com/TomNicholas/datatree

**Pros/ Cons**
- could overcome the issues of (2)
- not production ready
- metadata must be encoded in the key

**Comment by leabeusch** [comment]
@znicholls same comment as for #105 but probably even more important to get your input on this at some point!

**Comment by leabeusch** [comment]
I haven't understood everything going on here from a code perspective yet, so in case some of my statements don't make any sense, this could really just be because of a stupid misunderstanding -> sorry in advance for that. :sweat_smile: but I'm confident with some more time / explanations I'll catch up again eventually. :smile: I'll try to add a few things nevertheless already at this stage:


1. > IIRC @leabeusch would prefer to have one DataList for all variables. My preference would be to have one per variable. 

I just really want to keep the option for MESMER to go multivariate. Meaning: I'd like to be able to pass e.g., temperature and precip fields from a single ESM (but from various emission scenarios) at once to whatever forced response module (e.g., regression or sth fancier) & internal variability module (e.g., multivariate Gaussian after some transformations) I have available. Thus it seems weird to me to have those two variables in a different datalist. It would e.g., make more sense to have separate datalists per ESM rather than per variable to me. But maybe I have misunderstood what would need to be included in the same datalist to be able to achieve the functionality I describe?

2. This internal data structure would be applied to both our ESM runs and our emulations, correct?
3. If it really applies to emulations too: say we choose the datalist (which seems quite established by now ^^) -> how would you go about saving? Do you just save the individual dataarrays to netcdf? Would somehow include the metadata too? E.g., as attributes of the datarrays? Or just in the name of the file?

**Comment by mathause** [comment]
I had a chat with Sarah on her mesmer temperature/ precipitation analysis. She currently organizes her data in one large `xr.Dataset` (see below). It can be practical to have everything in one place. However, she currently only looks at one simulations (?) and what I am not entirely sure is how this will scale to several models, scenarios etc...


```
<xarray.Dataset>
Dimensions:                         (time: 1932, lat: 72, lon: 144, month: 12)
Coordinates:
  * time                            (time) object 1920-01-16 12:00:00 ... 208...
  * lon                             (lon) float32 -178.8 -176.2 ... 176.2 178.8
  * lat                             (lat) float32 -88.75 -86.25 ... 86.25 88.75
  * month                           (month) int64 1 2 3 4 5 6 7 8 9 10 11 12
Data variables:
    pr                              (time, lat, lon) float32 ...
    tas                             (time, lat, lon) float32 ...
    lsm                             (lat, lon) float32 nan nan nan ... nan nan
    log_pr                          (time, lat, lon) float32 ...
    monthly_global_tas              (time) float32 266.0 266.2 ... 275.8 271.3
    tas_hp                          (time, lat, lon) float32 ...
    pr_hp                           (time, lat, lon) float32 ...
    log_pr_hp                       (time, lat, lon) float32 ...
    tas_res_variability             (time, lat, lon) float32 ...
    log_pr_res_variability          (time, lat, lon) float32 ...
    tas_autoreg_res_variability     (time, lat, lon) float32 ...
    log_pr_autoreg_res_variability  (time, lat, lon) float32 ...
    tas_generated_variability       (time, lat, lon) float32 ...
    log_pr_generated_variability    (time, lat, lon) float32 ...
    tas_autoregressed               (time, lat, lon) float32 ...
    log_pr_autoregressed            (time, lat, lon) float32 ...
    tas_autoreg_slope               (month, lat, lon) float32 ...
    tas_autoreg_offset              (month, lat, lon) float32 ...
    log_pr_autoreg_slope            (month, lat, lon) float32 ...
    log_pr_autoreg_offset           (month, lat, lon) float32 ...
Attributes:
    CDI:          Climate Data Interface version 1.9.10 (https://mpimet.mpg.d...
    Conventions:  CF-1.6
    history:      Tue Oct 19 15:25:28 2021: cdo remapcon2,output_grid_descrip...
    CDO:          Climate Data Operators version 1.9.10 (https://mpimet.mpg.d...

```

**Comment by znicholls** [comment]
> I am not entirely sure is how this will scale to several models, scenarios etc

Yes, particularly how it would handle the very annoying history/scenario split

**Comment by mathause** [comment]
But it's of course super convenient because you only need to schlep one data structure around.

---

An alternative to a `DataList` would be to (ab-)use a pandas `DataFrame` - similar to geopandas. Either by subclassing or via an accessor (see [extending pandas](https://pandas.pydata.org/pandas-docs/stable/development/extending.html)). The `DataFrame` or `Dataset` would then need a dedicated column and could have an accessor to do fun stuff with it. But again not sure how this would work out exactly.

**Comment by mathause** [comment]
A pointer to cf_xarray and it's accessor that seems to wrap various methods and classes of xarray (as a potential inspiration how to do this for a `DataList` (or `xrDataList` or whatever I called it).

**Comment by mathause** [comment]
- DataTree is currently being ported into xarray. This effort has just started and is certainly still far off. When I tried it out I was quite confused how it's supposed to work but I maybe dis just not spend enough time with it (and there are prob. some rough edges that need polishing)
- DataList:
  - My sub-optimal implementation I used for the IPCC stuff can be found under [IPCC-WG1/Chapter-11 -> datalist.py](https://github.com/IPCC-WG1/Chapter-11/blob/main/code/utils/datalist.py). Check also my ideas for [improvements](https://github.com/IPCC-WG1/Chapter-11/tree/main/code#ideas-for-code-improvement)
  - I have started implementing a DataList at `/net/cfc/landclim1/mathause/projects/mesmer/datalist`

**Comment by znicholls** [comment]
I had a play with data tree and really liked it. For the kind of data we have here, it seemed the right choice. Looking at your links, putting some of the data list ideas onto data tree (or a class that wraps a data tree) seems the best choice to me...

filefinder looks sick by the way, can't wait to use that.

@mathause I am hoping to build some sort of CMIP data processing thing this year (we need it for MAGICC calibration). Would you be interested in collaborating or is finding time for this an ongoing dream that is never realised?

**Comment by mathause** [comment]
> I had a play with data tree and really liked it. For the kind of data we have here, it seemed the right choice (at least for container, then you just build on top of it as needed). Looking at your links, putting some of the data list ideas onto data tree (or a class that wraps a data tree) seems the best choice to me...

@veni-vidi-vici-dormivi (Victoria) started playing with datatree and it indeed looks promising

> filefinder looks sick by the way, can't wait to use that.

:blush: 
 
> @mathause I am hoping to build some sort of CMIP data processing thing this year (we need it for MAGICC calibration). Would you be interested in collaborating or is finding time for this an ongoing dream that is never realised?

Yes, please keep me in the loop! I assume you have heard of [jbusecke/xMIP](https://github.com/jbusecke/xMIP)?

**Comment by znicholls** [comment]
> Yes, please keep me in the loop! I assume you have heard of [jbusecke/xMIP](https://github.com/jbusecke/xMIP)?

Yep. We want to be able to download our own data though so our use of it depends a bit on how tightly coupled it is to the pangeo data archives.
